### DISCUSSION TITLE
Scalability

### DISCUSSION THREAD

**Discussion by ernst86913** [directus/directus#9781] [Feature Requests]
Hi - 
whats your view on scalability of Directus?

Read an interesting article about it:  https://softwareontheroad.com/nodejs-scalability-issues/

Are there any plans to add cluster mode? Or is this handled differently in Directus implementation?

Thx,
Roman

**Comment by rijkvanzanten** [comment]
Heya!

When I run Directus for a project, I tend to run Directus containerized, and horizontally scale based on the number of containers. This allows Directus to scale horizontally without any limits on the service itself (the limit will be based on the amount connections your database can handle, or how many targets your load balancer can have etc). By combining this with vertical scaling on the containers itself, the Node service should never really be the bottleneck. For anything stateful in Directus (like tracking rate limiters, file uploads, caches etc) there's configuration available to use shared managed resources, like Redis or S3/Azure/GCS storage etc. There's no plan to currently do any clustering through Worker nodes within the Node process itself, as reliant multi-threading is incredibly finicky to get just right within Node, and doesn't really offer any benefits that scaling on the infrastructure layer doesn't already provide 👍🏻

**Reply by nodeworks** [reply]
Is it possible to use cron job hooks when scaling horizontally? wouldn’t the jobs run at the same time and be redundant?

**Reply by azrikahar** [reply]
@nodeworks yes that's true, they will indeed be redundant in that case. Although there's no native support for it now, you should be able to use something like bullmq to prevent that from happening in your cron job hooks 👍

**Comment by ernst86913** [comment]
Hey Rijk,
Thanks for explanations… fully agree with how you are handling it… and that in this case vertical scaling does not make sense.

Are you running your projects usually using Kubernetes? Or „simple“ virtual docker containers?

LG,
Roman

Von meinem iPhone gesendet

**Reply by rijkvanzanten** [reply]
> Are you running your projects usually using Kubernetes? Or „simple“ virtual docker containers?

Really depends on the project's scope and other related infrastructure. I tend to use managed services like AWS' ECS, Azure's App Service, or DigitalOcean's App Platform rather than using a VPS with something like Kubernetes or Docker directly.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
